### PR TITLE
Add stop() to train loop when callback conditions are met

### PR DESCRIPTION
**Pull request by dhairyagandhi96**
This is with regard to the conversation on /FluxML/Flux.jl#131 for exposing a `stop()` function to break the training loop when a callback condition is met.

@MikeInnes

### CODE REVIEW

File: src/Flux.jl [hunk 1]
```diff
@@ -21,7 +21,7 @@ include("optimise/Optimise.jl")
 using .Optimise
 using .Optimise: @epochs
 export SGD, ADAM, ADAMW, AdaMax, Momentum, Nesterov,
-       RMSProp, ADAGrad, ADADelta, AMSGrad, NADAM
+       RMSProp, ADAGrad, ADADelta, AMSGrad, NADAM, stop, StopException
```
Comment (MikeInnes) [inline src/Flux.jl]: I think these names are a bit too generic for us to export them. Let's just have the API be `Flux.stop()`.

File: src/optimise/train.jl [hunk 2]
```diff
@@ -39,7 +58,16 @@ function train!(loss, data, opt; cb = () -> ())
     l = loss(d...)
     @interrupts back!(l)
     opt()
-    cb() == :stop && break
+    try
```
Comment (MikeInnes) [inline src/optimise/train.jl]: How about putting the try/catch around the model as well? Then you could stop from inside the loss function.

File: src/optimise/train.jl [hunk 3]
```diff
@@ -39,7 +58,16 @@ function train!(loss, data, opt; cb = () -> ())
...
+      cb()
```
Comment (MikeInnes) [inline src/optimise/train.jl]: Let's check for `:stop` here so that it's backwards compatible; we can emit a warning that it's deprecated.

File: src/optimise/train.jl [hunk 4]
```diff
@@ -39,7 +58,16 @@ function train!(loss, data, opt; cb = () -> ())
...
+    catch ex
+      if ex isa StopException
+        @info "Stop condition met"
```
Comment (MikeInnes) [inline src/optimise/train.jl]: I guess this is for debugging, but I don't think we need to include it.

File: src/Flux.jl [hunk 5]
```diff
@@ -23,6 +23,7 @@ using .Optimise: @epochs
 export SGD, ADAM, ADAMW, AdaMax, Momentum, Nesterov,
        RMSProp, ADAGrad, ADADelta, AMSGrad, NADAM
 
+
```
Comment (MikeInnes) [inline src/Flux.jl]: Best to avoid spurious whitespace changes like this (see also the deleted endlines further down, Juno will take care of those for you).

File: src/optimise/train.jl [hunk 6]
```diff
@@ -36,10 +55,18 @@ function train!(loss, data, opt; cb = () -> ())
...
   @progress for d in data
-    l = loss(d...)
-    @interrupts back!(l)
-    opt()
-    cb() == :stop && break
+    try
+      l = loss(d...)
+      @interrupts back!(l)
+      opt()
+      cb() == :stop && break
```
Comment (MikeInnes) [inline src/optimise/train.jl]: do add the depwarn in this case; we'll want to stop support this way of stopping in future.

**Comment by MikeInnes** [comment]
Perfect, thanks a lot @dhairyagandhi96!
